joule-labs/joule-extension
# [WIP] Implement allowances that auto-pay invoices

Closes #42

### Description

This adds a new module, `appconf` that can be used to add customizations to each webapp. The one implemented here is allowances, the idea of setting up rules that allow automatic payments to happen without user intervention.

This also adds a status bar that gives you quick actions you can take on the current page you're on.

A more thorough write-up of this PR is up on Medium: https://medium.com/p/2b08bec75e3a

### TODOs

* Make more chain agnostic (hard-coded use of `sats` in a few places)
* Find good places to provide fiat equivalents
* ???

### Steps to Test

1. Go to a WebLN page that you've enabled
2. Confirm you see the enabled status bar
3. Click the disable button, confirm it disables WebLN for app
4. Re-enable it, and click on the piggy bank, confirm it created a default allowance for you
5. Attempt to make a payment that fits the constraints, confirm it happened automatically for you
6. Attempt to make a payment that does _not_ fit the constraints, confirm you were prompted instead
7. Configure the settings and make sure they all work as expected
8. Add a new allowance directly from the allowances UI, confirm it correctly works when manually entered

## Screenshots

### Status Bar

### Allowance page

<img width="408" alt="Screen Shot 2019-06-06 at 5 42 29 PM" src="https://user-images.githubusercontent.com/649992/59449421-00a6e980-8dd5-11e9-8a51-1238376d7351.png">

### Notifications

<img width="363" alt="Screen Shot 2019-06-06 at 5 40 31 PM" src="https://user-images.githubusercontent.com/649992/59455031-5c776f80-8de1-11e9-9340-e00d16568881.png">

<img width="358" alt="Screen Shot 2019-06-06 at 5 40 38 PM" src="https://user-images.githubusercontent.com/649992/59455041-5f726000-8de1-11e9-96b8-99462dbc83ef.png">
